core/src/main/java/org/apache/iceberg/data/GenericRecord.java

```diff
@@ -65,13 +68,36 @@ private GenericRecord(StructType struct) {
     this.nameToPos = NAME_MAP_CACHE.get(struct);
   }
 
-  private GenericRecord(GenericRecord toCopy) {
+  private GenericRecord(GenericRecord toCopy, boolean deepCopyValues) {
```

Comment:
I don't think we should have a flag for this and I think we should have the code a bit more similar to the other classes where we implement this same logic for example

https://github.com/apache/iceberg/blob/fb02bd2d713c9bef8b8ccee334f6d65087c344b2/core/src/main/java/org/apache/iceberg/PartitionData.java#L227-L253

I'm not sure why we need to do conversions for the time types?